### PR TITLE
Add a ServiceTalk debugging utilities example

### PR DESCRIPTION
**Pull request by bondolo**
Motivation:
ServiceTalk has purpose-built features for improving the debugging
experience for both applications and for ServiceTalk itself. An example
with explanatory comments would improve exposure of these features.

Modifications:
A new HTTP debugging example is provided demonstrating wire logging,
disabling offloading, disabling async context.

Result:
ServiceTalk debugging features are more visible.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -30,6 +30,8 @@ if (!repositories) {
 apply plugin: "io.servicetalk.servicetalk-gradle-plugin-internal-root"
 
 task validateLocalDocSite(type: Exec) {
+  group 'Documentation'
```
Comment (bondolo) [inline build.gradle:33]: This change is unrelated to the debugging utilities except that it was needed to validate the documentation change and did not appear in the Documentation tasks list.

File: servicetalk-examples/http/debugging/src/main/java/io/servicetalk/examples/http/debugging/DebuggingExampleClient.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+/**
+ * The async "Hello World" example with debugging features enabled.
+ */
+public final class DebuggingExampleClient {
+
+    static {
+        // Enables visibility for all wire log messages
+        System.setProperty("servicetalk.logger.wireLogLevel", "TRACE");
+        // Disables the context associated with individual request/responses to simplify execution tracing
```
Comment (Scottmitch) [inline servicetalk-examples/http/debugging/src/main/java/io/servicetalk/examples/http/debugging/DebuggingExampleClient.java:38]: add a disclaimer/caveat that features which depend upon `AsyncContext` will stop working (e.g. tracing, auth, etc.)

File: servicetalk-examples/http/debugging/src/main/java/io/servicetalk/examples/http/debugging/DebuggingExampleServer.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        System.setProperty("servicetalk.logger.wireLogLevel", "TRACE");
+        // Disables the context associated with individual request/responses to simplify execution tracing
+        AsyncContext.disable();
+    }
+
+    public static void main(String... args) throws Exception {
+        HttpServers.forPort(8080)
+                // Disables asynchronous offloading to simplify execution tracing
+                .executionStrategy(HttpExecutionStrategies.noOffloadsStrategy())
+                // Enables detailed logging of I/O and I/O states.
```
Comment (Scottmitch) [inline servicetalk-examples/http/debugging/src/main/java/io/servicetalk/examples/http/debugging/DebuggingExampleServer.java:43]: is it worth also adding the frame logger for h2? `.protocols(h1Default(), h2().enableFrameLogging("h2-logging").build())` /cc @idelpivnitskiy

**Review by Scottmitch — APPROVED** [review]

File: servicetalk-examples/http/debugging/build.gradle [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+apply plugin: "java"
+apply from: "../../gradle/idea.gradle"
+
+dependencies {
+  implementation project(":servicetalk-annotations")
+  implementation project(":servicetalk-encoding-netty")
```
Comment (idelpivnitskiy) [inline servicetalk-examples/http/debugging/build.gradle:22]: This dependency is not used in this example

File: servicetalk-examples/docs/modules/ROOT/pages/http/index.adoc [hunk 5]
```diff
@@ -73,10 +73,22 @@ blocking iterable stream of buffers.
...
+receives a response greeting the posted name as a single content.
+
+[#Debugging]
+== Debugging
+
+Extends the async "Hello World" example to demonstrate some debugging features available
+ for debugging ServiceTalk applications.
+
+* link:{source-root}/servicetalk-examples/http/debugging/src/main/java/io/servicetalk/examples/http/debugging/DebuggingExampleServer.java[DebuggingExampleServer] - a server that demonstrates
+the asynchronous API and responds with a simple `Hello World!` response body, optionally customized for a specific name for `POST` requests, as a `text/plain`.
```
Comment (idelpivnitskiy) [inline servicetalk-examples/docs/modules/ROOT/pages/http/index.adoc:89]: Consider clarifying the real purpose of this example - it demonstrates debuggability features, it doesn't matter what API it uses or what type or request sends.

File: servicetalk-examples/http/debugging/src/main/java/io/servicetalk/examples/http/debugging/DebuggingExampleClient.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+                // Disables asynchronous offloading to simplify execution tracing
+                .executionStrategy(HttpExecutionStrategies.noOffloadsStrategy())
+                // Enables detailed logging of I/O and I/O states.
+                // Be sure to also enable the logger in your logging config file (log4j2.xml for this example)
+                .enableWireLogging("servicetalk-examples-wire-logger", TRACE, Boolean.TRUE::booleanValue)
+                .build()) {
+            // This example is demonstrating asynchronous execution, but needs to prevent the main thread from exiting
+            // before the response has been processed. This isn't typical usage for a streaming API but is useful for
+            // demonstration purposes.
+            CountDownLatch responseProcessedLatch = new CountDownLatch(2);
```
Comment (idelpivnitskiy) [inline servicetalk-examples/http/debugging/src/main/java/io/servicetalk/examples/http/debugging/DebuggingExampleClient.java:53]: Will it be better to use `BlockingHttpClient` to simplify the example and make a better focus on debuggability features?

**Review by idelpivnitskiy — COMMENTED** [review]
Good idea to demonstrate this features! Have a few suggestions to improve the focus:

File: servicetalk-examples/http/debugging/src/main/java/io/servicetalk/examples/http/debugging/DebuggingExampleClient.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+/**
+ * The async "Hello World" example with debugging features enabled.
+ */
+public final class DebuggingExampleClient {
+
+    static {
+        // Enables visibility for all wire log messages
+        System.setProperty("servicetalk.logger.wireLogLevel", "TRACE");
+        // Disables the context associated with individual request/responses to simplify execution tracing
+        AsyncContext.disable();
```
Comment (idelpivnitskiy) [inline servicetalk-examples/http/debugging/src/main/java/io/servicetalk/examples/http/debugging/DebuggingExampleClient.java:39]: Because it's easy to forget to remove this statement after the debugging we may demonstrate it can be protected by the system property. Something like:
```java
static {
    if (Boolean.getBoolean("debuggingMode")) {
        AsyncContext.disable();
    }
}
```

File: servicetalk-examples/http/debugging/src/main/java/io/servicetalk/examples/http/debugging/DebuggingExampleClient.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+import static io.servicetalk.logging.api.LogLevel.TRACE;
+
+/**
+ * The async "Hello World" example with debugging features enabled.
+ */
+public final class DebuggingExampleClient {
+
+    static {
+        // Enables visibility for all wire log messages
+        System.setProperty("servicetalk.logger.wireLogLevel", "TRACE");
```
[the remark this answers is not in the document]
Comment (bondolo) [reply servicetalk-examples/http/debugging/src/main/java/io/servicetalk/examples/http/debugging/DebuggingExampleClient.java:122]: OK I will remove it.

File: servicetalk-examples/http/debugging/src/main/java/io/servicetalk/examples/http/debugging/DebuggingExampleClient.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        // Disables the context associated with individual request/responses to simplify execution tracing
+        AsyncContext.disable();
+    }
+
+    public static void main(String... args) throws Exception {
+        try (HttpClient client = HttpClients.forSingleAddress("localhost", 8080)
+                // Disables asynchronous offloading to simplify execution tracing
+                .executionStrategy(HttpExecutionStrategies.noOffloadsStrategy())
```
[the remark this answers is not in the document]
Comment (bondolo) [reply servicetalk-examples/http/debugging/src/main/java/io/servicetalk/examples/http/debugging/DebuggingExampleClient.java:130]: Good clarification. I have marked it optional, added additional explanation and commented it out by default. It was useful for us to debug across a boundary that would have normally involved an async thread handoff.

File: servicetalk-examples/http/debugging/build.gradle (hunk 4 above)
Comment (bondolo) [reply servicetalk-examples/http/debugging/build.gradle:22]: Is there an IntelliJ detector for this? I used to use one in Netbeans with Maven and would like to turn it on for Gradle if I could.

File: servicetalk-examples/http/debugging/src/main/java/io/servicetalk/examples/http/debugging/DebuggingExampleClient.java (hunk 7 above)
Comment (bondolo) [reply servicetalk-examples/http/debugging/src/main/java/io/servicetalk/examples/http/debugging/DebuggingExampleClient.java]: It would be good to standardize how we use system properties and/or environment variables for configuration and control of debugging features. It is definitely a useful tool. In other orgs I have used 

`(public|private)` static final boolean SOMETHING_DEBUG = Boolean.getBoolean("package.name.Feature.debug");`

and then triggered all use off of the boolean. One advantage is that it lets the JIT remove dead code and also makes it easy to find all the references.

**Comment by bondolo** [comment]
Thank you for the feedback. I think this will be a really useful example.

File: servicetalk-examples/http/debugging/build.gradle (hunk 4 above)
Comment (idelpivnitskiy) [reply servicetalk-examples/http/debugging/build.gradle]: Not that I'm aware of 🤷‍♂️
